### ISSUE TITLE
Deployment

### ISSUE THREAD

**Issue by michalk8** [msmdev/pyGPCCA#1]
Before we put this on PyPI/conda-forge, here's a list of TODOs:

- [x] create README (also add small code sample)  #2 
- [x] create comprehensive installation intructions for Linux/macOS  #2 
- [x] add tests for stationary_distribution #3 
- [x] use `k=m` in https://github.com/msmdev/pyGPCCA/blob/main/pygpcca/_sorted_schur.py#L358 - correct @msmdev ?
- [x] check for typos in the docs/fix badge links
- [x] add PYPI_PASSWORD repo secret

After this, we can put in on PyPI and afterwards, we can create a PR to conda-forge.

**Comment by msmdev** [comment]
One thing that came to my mind: how will we handle versioning?

**Comment by michalk8** [comment]
> One thing that came to my mind: how will we handle versioning?

There's https://github.com/msmdev/pyGPCCA/blob/main/.bumpversion.cfg which creates the bump commit (e.g. `bump2version patch`) + creates a new tag. You just need to push the commit with the tag to the repo (I have in my git config `push.followtags=true` so that they are pushed simultaneously). Then CI runs and if it's OK, it will create new build on PyPI.

**Comment by Marius1311** [comment]
Okay!

**Comment by msmdev** [comment]
I updated the README (currently just on the update-readme branch).
I think we are very close to be able to deploy (and should soon, since pyGPCCA should be ready to use in cellrank until the re-submission of the paper) - so what still needs to be done and who will take care of what?

**Comment by Marius1311** [comment]
Let me know if I can help in any way - I'm pretty occupied with the revision at the moment and trying to focus on that, but could take over little tasks if needed. 

One thing that would be nice is a concept figure in the README - I think repos with a concept figure just look much more appealing and friendly.

**Comment by msmdev** [comment]
A concept figure is a very nice idea! Do you have anything in mind?

**Comment by Marius1311** [comment]
Mhm, is there anything in your old papers we could use?

**Comment by msmdev** [comment]
Hmmm, currently only an old graphical abstract comes to my mind, but I fear it's is to specific...
![Graphical_Abstract3_scaled](https://user-images.githubusercontent.com/43750470/105192075-0aa7d700-5b38-11eb-924b-42cf7e88f70d.png)

**Comment by Marius1311** [comment]
yes, this is too specific to protein folding, we want something more general that focuses on GPCCA rather than an application. I will see whether I can come up with something. However, this is really not essentially, we can put in on pypi without this and add later.

**Comment by msmdev** [comment]
@michalk8 IMPORTANT: even if we are ready soon, we have to wait with the deployment on PyPI and conda until I have an OK from Berlin... (this should come today/tomorrow)

**Comment by michalk8** [comment]
Ok, sounds good. Here are the deployment steps you need do:

0. make sure you have `push.followtags=true` in your git config
1. make sure you do `pip install -e'.[dev]'`
1.5 `git checkout main`
2. run `bump2version <part>` where `<part>` is one of `major`, `minor`, `patch` (as in `{major.minor.patch}`)*
2.1 sanity check: verify that new tag has been created: `git tag`
2.2. verify that a new commit with msg like  `Bump version: 1.0.0 → 1.1.0` has been created
2.3 super extra sanity check: verify the new version has been written in these 3 files: `setup.py`, `pygpcca/__init__`
3. run `git push`
4. let CI run, if the linting/tests are all OK, deployment starts (you can watch it go here: https://github.com/msmdev/pyGPCCA/actions)
5. assuming everything is correct (CI conditions/PyPI credentials), it should be on PyPI

If anything goes awry during commiting, don't push the commit, just revert it (don't forget to also delete the newly created tag [if it was created] as `git tag -d <tag_id>`.
Locally, I've tried `bump2version major`, worked fine (the new version is `1.0.0`, which is expected). If you're also trying this out, you might want to try `--dry-run --verbose` to see what would be changed.
If you don't want to create a new tag, just specify `--no-tag`. No tags push to main = no new release.

*alternative is to specify `--new-version {major}.{minor}.patch`

**Comment by msmdev** [comment]
@michalk8 ok and thank you for the detailed guidance!
I will wait for the input regarding credits from @Marius1311 before deploying.
Just one question: I would preferably go for `bump2version major` in step 2. - correct?

**Comment by michalk8** [comment]
>  Just one question: I would preferably go for bump2version major in step 2. - correct?

Yes. Also, we need to change the development status classifiers on PyPI: https://github.com/msmdev/pyGPCCA/blob/main/setup.py#L53 - should be `Development Status :: 5 - Production/Stable`.

Also, if you find any other useful classifier in https://pypi.org/classifiers/, feel free to add. Actually, you can do that in 1 commit when bumping the version as `bump2version major --allow-dirty` (but try with `--dry-run` first).

**Comment by msmdev** [comment]
@michalk8 I also registered that the coverage is just around 76%.
How about that?

**Comment by michalk8** [comment]
@msmdev if you think any lines that we miss here is worth testing for, then let's create a separate issue for that

https://codecov.io/gh/msmdev/pyGPCCA/tree/d5e589ca168684e16f52c14f25a8a86fdcebb6c1/pygpcca (this is the last non-skipped commit on Ubuntu, Python3.8 on `add-example` branch).

The `_checks.py` file has ~47% coverage, but most of the stuff we're missing are corner cases (ValueError, etc.).

**Comment by msmdev** [comment]
Ok, lets leave it for the time being - can be handled in a future version.

**Comment by msmdev** [comment]
@michalk8 another question: I thought that we originally enforced the data type of the input transition matrix to be double and also ensured that all floating point calculations were double precision, but I couldn't it - is it still done? Or are we currently just trusting in python/numpy internal type handling here?

**Comment by Marius1311** [comment]
Thanks guys for pushing this, and sorry for not being super active at the moment, I'm quite occupied with revision work.

**Comment by michalk8** [comment]
Binder now works as expected.

**Comment by msmdev** [comment]
> Ok, sounds good. Here are the deployment steps you need do:
> 
> 1. make sure you have `push.followtags=true` in your git config
> 2. make sure you do `pip install -e'.[dev]'`
>    1.5 `git checkout main`
> 3. run `bump2version <part>` where `<part>` is one of `major`, `minor`, `patch` (as in `{major.minor.patch}`)*
>    2.1 sanity check: verify that new tag has been created: `git tag`
>    2.2. verify that a new commit with msg like  `Bump version: 1.0.0 → 1.1.0` has been created
>    2.3 super extra sanity check: verify the new version has been written in these 3 files: `setup.py`, `pygpcca/__init__`
> 4. run `git push`
> 5. let CI run, if the linting/tests are all OK, deployment starts (you can watch it go here: https://github.com/msmdev/pyGPCCA/actions)
> 6. assuming everything is correct (CI conditions/PyPI credentials), it should be on PyPI
> 
> If anything goes awry during commiting, don't push the commit, just revert it (don't forget to also delete the newly created tag [if it was created] as `git tag -d <tag_id>`.
> Locally, I've tried `bump2version major`, worked fine (the new version is `1.0.0`, which is expected). If you're also trying this out, you might want to try `--dry-run --verbose` to see what would be changed.
> If you don't want to create a new tag, just specify `--no-tag`. No tags push to main = no new release.
> 
> *alternative is to specify `--new-version {major}.{minor}.patch`

@michalk8 Regarding 1.: Is it done via 
`git config --global push.followTags true`?

**Comment by michalk8** [comment]
> @michalk8 Regarding 1.: Is it done via
git config --global push.followTags true?

Yes.

**Comment by msmdev** [comment]
@michalk8 somehow all tests passed, but deploy was first skipped and just started when I clicked it, but failed [here](https://github.com/msmdev/pyGPCCA/runs/1793655220?check_suite_focus=true)
resp. was skipped [here](https://github.com/msmdev/pyGPCCA/actions/runs/521647335)

**Comment by michalk8** [comment]
@msmdev it cannot build wheel for some reason. Looking closer at pyproject.toml, it isn't specified as a requirement in the `build-system`. Can I revert the last commit + delete the tag on upstream and fix this? Then you can again run `bump2version`.

**Comment by msmdev** [comment]
@michalk8 ok!

**Comment by michalk8** [comment]
@msmdev done in https://github.com/msmdev/pyGPCCA/commit/4fac09ffc514354855398cfa98381a93e6f15af0. Make sure you pull from main and locally delete your tag locally as well `git tag -d v1.0.0`, before running `bump2version` again.

**Comment by msmdev** [comment]
@michalk8 should I mb. additionally update anything locally, e.g. `pip` etc.?

**Comment by michalk8** [comment]
Not really, if you want, you can double-check (I already did):
```bash
pip install build
python -m build --sdist --wheel --outdir dist/
```
and it should  have 0 exit status.

**Comment by msmdev** [comment]
@michalk8 after `git pull origin main` I still have the bump commit in my log:
```
a6125b5 (HEAD -> main) Merge branch 'main' of https://github.com/msmdev/pyGPCCA into main
4fac09f (origin/main, origin/HEAD) Add forgotten 'wheel' in build-requirements
6a44fab (tag: v1.0.0) Bump version: 0.0.0 → 1.0.0
7006afb Improve `GPCCA.optimize()`: Change complex conjugate splitting handling etc (#13)
073cffb Fix rtd (#16)
```

**Comment by michalk8** [comment]
Just try `git reset --hard 073cffb` and `git pull` should do the trick.

**Comment by msmdev** [comment]
Jup, worked.

**Comment by msmdev** [comment]
Will remove the tag now and proceed...

**Comment by msmdev** [comment]
> ```shell
> pip install build
> python -m build --sdist --wheel --outdir dist/
> ```

Ok, worked it seems.

**Comment by michalk8** [comment]
> Ok, worked it seems.

Great. Let's see now if the PyPI upload also works.

**Comment by msmdev** [comment]
`bump2version` worked, tag 1.0.0 was created, commit is also there and in setup.py it is written.
Will push now.

**Comment by msmdev** [comment]
@michalk8 when the tests passed, where should I first click deploy in [actions](https://github.com/msmdev/pyGPCCA/actions)?:
* Bump version: 0.0.0 → 1.0.0, CI #107, [main](https://github.com/msmdev/pyGPCCA/actions/runs/521750544)
* Bump version: 0.0.0 → 1.0.0, CI #108, [v.1.0.0](https://github.com/msmdev/pyGPCCA/actions/runs/521750550)

**Comment by michalk8** [comment]
The `v.1.0.0`, since this one was triggered on `tag` push.

**Comment by msmdev** [comment]
ok!

**Comment by michalk8** [comment]
I've just watched it, the problem seems to be the README and the `..raw` directive in `.rst`.

**Comment by msmdev** [comment]
Yep, guessed the same x.x

**Comment by michalk8** [comment]
I will fix this short + 1 more issue, will let you know once it's ready.

**Comment by msmdev** [comment]
So, same procedure as before? you revert and fix and I try again?

**Comment by michalk8** [comment]
Yes, that's the plan. We have the check for this in tox as `tox -e readme`, but I forgot to add it to linting.

**Comment by msmdev** [comment]
Okdok

**Comment by michalk8** [comment]
@msmdev done, you can try again, hopefully for the last time.
One thing to be careful about: in [setup.py](https://github.com/msmdev/pyGPCCA/blob/main/setup.py#L21), we automatically set the description + long_descript from README (the description is line 35, the long_description is the full file, except for the `.. raw`, which is removed). For the long_description, I replace `|br|` with `\`, which should be the markdown line break.

**Comment by msmdev** [comment]
@michalk8 ok, I repeated everything and pushed - thumbs crossed.

**Comment by msmdev** [comment]
@michalk8 oook, everything worked this times besides of invalid authentication info...

**Comment by michalk8** [comment]
Yes, I've seen it - the username should be `__token__` as seen [here](https://github.com/msmdev/pyGPCCA/blob/main/.github/workflows/ci.yml#L171) and the password should be a repo secret named `PYPI_PASSWORD`.
Is this set up correctly? Here's how mine looks like:
![pypi](https://user-images.githubusercontent.com/46717574/106322685-81e71480-6276-11eb-8171-0a9cdd3a3979.png)

**Comment by msmdev** [comment]
> Yes, I've seen it - the username should be `__token__` as seen [here](https://github.com/msmdev/pyGPCCA/blob/main/.github/workflows/ci.yml#L171) and the password should be a repo secret named `PYPI_PASSWORD`.
> Is this set up correctly? Here's how mine looks like:
> ![pypi](https://user-images.githubusercontent.com/46717574/106322685-81e71480-6276-11eb-8171-0a9cdd3a3979.png)

hmmm, two things:
* I have set up tow-factor auth
* my token is named `PYPI_PASSWORD` and not `__token__`

**Comment by michalk8** [comment]
> my token is named PYPI_PASSWORD and not __token__

This is the issue most likely, we can either rename the user in the CI to also `PYPI_PASSWORD` or change token name in your `PyPI`.

**Comment by msmdev** [comment]
@michalk8 I will delete and name to `__token__` on PyPI and add this on GitHub.

**Comment by michalk8** [comment]
Ok, in the meantime, I will revert the `bump2version` commit.

**Comment by michalk8** [comment]
(the CI should run faster now, since it's cached - when I modified `tox.ini`, new cache was created).

**Comment by michalk8** [comment]
> Ok, in the meantime, I will revert the bump2version commit.

Done.

**Comment by msmdev** [comment]
Even the name is __token__ now in PyPI should it still be named PYPI_PASSWORD in GitHub secrets?

**Comment by michalk8** [comment]
> @michalk8 Even the name is __token__ now in PyPI should it still be named PYPI_PASSWORD in GitHub secrets?

Yes. In `PyPI`, token name should be `__name__`. The value of the token, should be saved in this repo's secrets
with the name `PYPI_PASSWORD`.

**Comment by msmdev** [comment]
Ok, then it should be alright now - will push now.

**Comment by msmdev** [comment]
@michalk8 ok funny, now it fails because of my dummy email...
you can insert a real one there: bernhard-reuter@gmx.de

**Comment by michalk8** [comment]
Done, you can try again.

**Comment by michalk8** [comment]
@msmdev Congratulations: https://pypi.org/project/pygpcca/1.0.0/ !

**Comment by michalk8** [comment]
The markdown (\) is not properly rendered, but that can be fixed.

**Comment by msmdev** [comment]
@michalk8 its fine for now...
Cheeers, it is deployed!
I owe you a bottle of champagne or some of my self-brewed beer (yes, I am also a home brewer) :-D

**Comment by michalk8** [comment]
@msmdev cheers! The home-brewed beer sounds really nice!

**Comment by msmdev** [comment]
@michalk8 if I would like to make some minor style fixes, like fixing the markdown (\), will this require a new version tag and everything?

**Comment by michalk8** [comment]
For it to be visible on https://pypi.org/project/pygpcca/, yes.

**Comment by michalk8** [comment]
I've found the recipe for conda (though it was lost):
```yml
{% set name = "pyGPCCA" %}
{% set version = "1.0.0" %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
  sha256: # TODO

build:
  number: 0
  noarch: python
  script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv"
  skip: True  # [py<36 and not win]

requirements:
  host:
    - pip
    - python
    - setuptools_scm
  run:
    - python
    - numpy >=1.15.0
    - scipy >=1.5.0
    - docrep >=0.3.1
    - mpi4py >=3.0.3
    - petsc >=3.13.0,!=3.14.0
    - slepc >=3.13.0
    - petsc4py >=3.13.0,!=3.14.0
    - slepc4py >=3.13.0

test:
  imports:
    - pygpcca
  commands:
    - pytest -vv
  requires:
    - pytest
    - pytest-mock

about:
  home: https://github.com/msmdev/pygpcca
  license: LGPL-3.0-or-later
  license_family: LGPL
  license_file: LICENSE.txt
  summary: 'Generalized Perron Cluster Cluster Analysis'
  description: |
    # TODO
  doc_url: https://pygpcca.readthedocs.io/en/latest
  dev_url: https://github.com/msmdev/pygpcca

extra:
  recipe-maintainers:
    - msmdev
```
Could you please fill-in the description?

**Comment by msmdev** [comment]
@michalk8 cool - I will fill in the description tomorrow :)

**Comment by msmdev** [comment]
> The markdown (\) is not properly rendered, but that can be fixed.

Couldn't find out how to get it right -.-

**Comment by Marius1311** [comment]
Amazing guys, well done, and thank you so much for both of your combined efforts! Could we add @michalk8 and my name somewhere on the Pypi page, either under "Acknowledgements" or in the Maintainers section? We both have pypi accounts, mine is https://pypi.org/user/marius1311/ and Mike's is https://pypi.org/user/michalk8/

**Comment by msmdev** [comment]
> Amazing guys, well done, and thank you so much for both of your combined efforts! Could we add @michalk8 and my name somewhere on the Pypi page, either under "Acknowledgements" or in the Maintainers section? We both have pypi accounts, mine is https://pypi.org/user/marius1311/ and Mike's is https://pypi.org/user/michalk8/

@Marius1311 @michalk8 I added a key contributors section in the README.rst.
This will be reflected on PyPI after the next release.

**Comment by msmdev** [comment]
> > The markdown () is not properly rendered, but that can be fixed.
> 
> Couldn't find out how to get it right -.-

@michalk8 any ideas how to get the markdown right on pypi? Simply `\n`?

**Comment by Marius1311** [comment]
> > Amazing guys, well done, and thank you so much for both of your combined efforts! Could we add @michalk8 and my name somewhere on the Pypi page, either under "Acknowledgements" or in the Maintainers section? We both have pypi accounts, mine is https://pypi.org/user/marius1311/ and Mike's is https://pypi.org/user/michalk8/
> 
> @Marius1311 @michalk8 I added a key contributors section in the README.rst.
> This will be reflected on PyPI after the next release.

wonderful, thanks!

**Comment by Marius1311** [comment]
Small point: I would add a link to the readthedocs to the 'about' section on github - people will be looking for this first.

**Comment by msmdev** [comment]
> Small point: I would add a link to the readthedocs to the 'about' section on github - people will be looking for this first.

@Marius1311 good point - thx!

**Comment by michalk8** [comment]
> @michalk8 any ideas how to get the markdown right on pypi? Simply \n?

I think it would be `<br/>`, but I would just change [here](https://github.com/msmdev/pyGPCCA/blob/main/setup.py#L22)
the `.replace('|br|', '\\')` into `.replace('|br|', '\n')`.

**Comment by msmdev** [comment]
> I've found the recipe for conda (though it was lost):
> 
> ```yaml
> {% set name = "pyGPCCA" %}
> {% set version = "1.0.0" %}
> 
> package:
>   name: {{ name|lower }}
>   version: {{ version }}
> 
> source:
>   url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
>   sha256: # TODO
> 
> build:
>   number: 0
>   noarch: python
>   script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv"
>   skip: True  # [py<36 and not win]
> 
> requirements:
>   host:
>     - pip
>     - python
>     - setuptools_scm
>   run:
>     - python
>     - numpy >=1.15.0
>     - scipy >=1.5.0
>     - docrep >=0.3.1
>     - mpi4py >=3.0.3
>     - petsc >=3.13.0,!=3.14.0
>     - slepc >=3.13.0
>     - petsc4py >=3.13.0,!=3.14.0
>     - slepc4py >=3.13.0
> 
> test:
>   imports:
>     - pygpcca
>   commands:
>     - pytest -vv
>   requires:
>     - pytest
>     - pytest-mock
> 
> about:
>   home: https://github.com/msmdev/pygpcca
>   license: LGPL-3.0-or-later
>   license_family: LGPL
>   license_file: LICENSE.txt
>   summary: 'Generalized Perron Cluster Cluster Analysis'
>   description: |
>     # TODO
>   doc_url: https://pygpcca.readthedocs.io/en/latest
>   dev_url: https://github.com/msmdev/pygpcca
> 
> extra:
>   recipe-maintainers:
>     - msmdev
> ```
> 
> Could you please fill-in the description? Once the description is done, I will get the sha256 from PyPI and we can open a PR at `conda-forge`. I will have to double check if the `skip` condition is right (think we don't need the `not win` [Windows]).

@michalk8 how about using the description section of the readme as description:

'Generalized Perron Cluster Cluster Analysis program to coarse-grain reversible and non-reversible Markov State Models.
Markov State Models (MSM) enable the identification and analysis of metastable states and related kinetics in a very instructive manner. They are widely used, e.g. to model molecular or cellular kinetics.
Common state-of-the-art Markov state modeling methods and tools are very well suited to model reversible processes in closed equilibrium systems. However, most are not well suited to deal with non-reversible or even non-autonomous processes of non-equilibrium systems.
To overcome this limitation, the Generalized Robust Perron Cluster Cluster Analysis (G-PCCA) was developed. The G-PCCA method implemented in the pyGPCCA program readily handles equilibrium as well as non-equilibrium data by utilizing real Schur vectors instead of eigenvectors.
pyGPCCA enables the semiautomatic coarse-graining of transition matrices representing the dynamics of the system under study. Utilizing pyGPCCA, metastable states as well as cyclic kinetics can be identified and modeled.
If you use pyGPCCA or parts of it, please cite Reuter 2018 (https://doi.org/10.1021/acs.jctc.8b00079).'

Or is this too long?

**Comment by msmdev** [comment]
> > @michalk8 any ideas how to get the markdown right on pypi? Simply \n?
> 
> I think it would be `<br/>`, but I would just change [here](https://github.com/msmdev/pyGPCCA/blob/main/setup.py#L22)
> the `.replace('|br|', '\\')` into `.replace('|br|', '\n')`.

Alright! Thx @michalk8.

**Comment by msmdev** [comment]
> > > @michalk8 any ideas how to get the markdown right on pypi? Simply \n?
> > 
> > 
> > I think it would be `<br/>`, but I would just change [here](https://github.com/msmdev/pyGPCCA/blob/main/setup.py#L22)
> > the `.replace('|br|', '\\')` into `.replace('|br|', '\n')`.
> 
> Alright! Thx @michalk8.

@michalk8 actually it is currently 
`.replace("|br|", r"\\")`
so
`.replace("|br|", r"\n")`?

**Comment by michalk8** [comment]
> Or is this too long?

I would keep it short, this is what we have in [CellRank](https://anaconda.org/bioconda/cellrank) - many packages
don't have any description (i.e. numpy/tensorflow...). Maybe just keep the first paragraph.

>  .replace("|br|", r"\n")?

It should be `.replace('|br|', '\n')`, without the raw strings.

**Comment by msmdev** [comment]
Ok, will think about something short.

**Comment by msmdev** [comment]
'pyGPCCA employs Generalized Perron Cluster Cluster Analysis (G-PCCA) to coarse-grain reversible and in particular non-reversible Markov State Models.
If you use pyGPCCA or parts of it, please cite Reuter et al. 2018 (https://doi.org/10.1021/acs.jctc.8b00079).'

**Comment by michalk8** [comment]
Sounds good, will make a PR to conda-later today and ping you in the related thread.

**Comment by msmdev** [comment]
@michalk8 guess I will make a new release tag now, since the markdown should be fixed now and deploy it to pypi.

**Comment by michalk8** [comment]
Sounds good, that way we use to correct hash for conda.

**Comment by msmdev** [comment]
@michalk8 v1.0.1 is deployed :)

**Comment by Marius1311** [comment]
Thanks for including the contributions @msmdev, it looks very nice!

**Comment by msmdev** [comment]
Since it the package is successfully deployed on pypi and conda-forge, I'm closing this now. :) Thx, again to @Marius1311 and especially to @michalk8!

**Comment by Marius1311** [comment]
Amazing work @msmdev and @michalk8 !
